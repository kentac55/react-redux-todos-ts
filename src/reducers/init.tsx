import { reducerWithInitialState } from 'typescript-fsa-reducers'
import { initAction } from '../actions'
import { InitState } from '../containers/init'

export const initReducer = reducerWithInitialState({
  loading: false,
  loaded: false,
  error: null,
} as InitState)
  .case(initAction.started, state => {
    return { ...state, loading: true }
  })
  .case(initAction.done, state => {
    return { ...state, loading: false, loaded: true }
  })
  .case(initAction.failed, (_, { error }) => {
    return { loading: false, loaded: true, error: error }
  })
