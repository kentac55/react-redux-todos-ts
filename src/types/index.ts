export type Todo = {
  id: number
  completed: boolean
  text: string
}

export enum VisibilityFilterKinds {
  ALL = 'SHOW_ALL',
  COMPLETED = 'SHOW_COMPLETED',
  ACTIVE = 'SHOW_ACTIVE',
}
