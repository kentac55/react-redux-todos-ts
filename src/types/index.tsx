export type Todo = {
  id: number
} & NewTodo

export type NewTodo = {
  completed: boolean
  text: string
}

export enum VisibilityFilterKinds {
  ALL = 'SHOW_ALL',
  COMPLETED = 'SHOW_COMPLETED',
  ACTIVE = 'SHOW_ACTIVE',
}
