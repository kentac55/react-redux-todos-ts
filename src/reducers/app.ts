import { reducerWithInitialState } from 'typescript-fsa-reducers'
import { addTodo, initApp, getTodo, getTodos, toggleTodo } from '../actions'
import { AsyncOpResult } from '../types'

export type AppState = {
  loading: boolean
  loaded: boolean
  error: {
    type: string
    message: string | null
  } | null
}

const failedHandler = (_: unknown, { error }: { error: Error }): AppState => {
  return {
    loaded: true,
    loading: false,
    error: { type: error.name, message: error.message },
  }
}
const doneHandler = (
  state: AppState,
  { result }: { result: AsyncOpResult<unknown> }
): AppState => {
  if (result.type === 'Ok') {
    return {
      ...state,
      loaded: true,
      loading: false,
    }
  } else {
    return {
      loaded: true,
      loading: false,
      error: { type: result.type, message: result.message },
    }
  }
}
const startedHandler = (state: AppState): AppState => {
  return { ...state, loading: true }
}

export const appReducer = reducerWithInitialState({
  loaded: false,
  loading: false,
  error: null,
} as AppState)
  .cases(
    [addTodo.started, getTodo.started, initApp.started, toggleTodo.started],
    startedHandler
  )
  .case(getTodos.started, startedHandler)
  .cases(
    [addTodo.failed, getTodo.failed, initApp.failed, toggleTodo.failed],
    failedHandler
  )
  .case(getTodos.failed, failedHandler)
  .cases(
    [addTodo.done, getTodo.done, initApp.done, toggleTodo.done],
    doneHandler
  )
  .case(getTodos.done, doneHandler)
