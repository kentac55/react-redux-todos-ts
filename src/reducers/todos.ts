import { reducerWithInitialState } from 'typescript-fsa-reducers'
import { Todo } from '../types'
import { getId } from '../repository'
import { addTodoAction, toggleTodoAction } from '../actions'

export const todoReducer = reducerWithInitialState([] as Todo[])
  .case(addTodoAction, (state, text) => {
    return [...state, { id: getId(), text, completed: false }]
  })
  .case(toggleTodoAction, (state, id) => {
    return state.map(todo => {
      return todo.id === id
        ? {
            ...todo,
            completed: !todo.completed,
          }
        : todo
    })
  })
