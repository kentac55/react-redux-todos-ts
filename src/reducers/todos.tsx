import { Reducer } from 'redux'
import { Todo, TodoAction } from '../types'

const todos: Reducer<Todo[], TodoAction> = (
  state: Todo[] = [],
  action: TodoAction
): Todo[] => {
  switch (action.type) {
    case 'ADD':
      return [
        ...state,
        {
          id: action.id,
          text: action.text,
          completed: false,
        },
      ]
    case 'TOGGLE':
      return state.map(
        (todo: Todo): Todo => {
          return todo.id === action.id
            ? {
                ...todo,
                completed: !todo.completed,
              }
            : todo
        }
      )
    default:
      return state
  }
}

export default todos
