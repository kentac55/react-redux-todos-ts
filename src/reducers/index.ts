import { combineReducers } from 'redux'
import { useSelector, TypedUseSelectorHook } from 'react-redux'
import { todoReducer } from './todos'
import { appReducer } from './app'
import { visibilityFilterReducer } from './visibilityFilters'

export const rootReducer = combineReducers({
  app: appReducer,
  todos: todoReducer,
  visibilityFilter: visibilityFilterReducer,
})

export type RootState = ReturnType<typeof rootReducer>

export const useTypedSelector: TypedUseSelectorHook<RootState> = useSelector
