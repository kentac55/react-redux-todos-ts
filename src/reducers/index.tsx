import { combineReducers } from 'redux'
import { todoReducer } from './todos'
import { visibilityFilterReducer } from './visibilityFilters'
import { initReducer } from './init'

export default combineReducers({
  init: initReducer,
  todos: todoReducer,
  visibilityFilter: visibilityFilterReducer,
})
