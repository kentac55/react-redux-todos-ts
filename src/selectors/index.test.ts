import { createStore } from 'redux'
import { appSelector, filterSelector, todoSelector } from '.'
import { VisibilityFilter } from '../types'
import { rootReducer } from './../reducers/index'

const storeState = createStore(rootReducer).getState()

it('should return application state', () => {
  expect(appSelector(storeState)).toMatchSnapshot()
})

it('should return filter state', () => {
  expect(filterSelector(storeState)).toMatchSnapshot()
})

it('should return todo state', () => {
  expect(todoSelector(storeState)).toMatchSnapshot()
})

const todos = [
  { id: 0, text: 'hoge', completed: false },
  { id: 1, text: 'fuga', completed: true },
  { id: 2, text: 'piyo', completed: false },
]
const todoState = { todos }
type VisibilityFilterState = {
  visibilityFilter: VisibilityFilter
}

it('should return all todos', () => {
  const visibilityFilterState: VisibilityFilterState = {
    visibilityFilter: 'SHOW_ALL',
  }

  expect(todoSelector.resultFunc(todoState, visibilityFilterState)).toBe(todos)
})

it('shoudl return completed todos', () => {
  const visibilityFilterState: VisibilityFilterState = {
    visibilityFilter: 'SHOW_COMPLETED',
  }
  expect(
    todoSelector.resultFunc(todoState, visibilityFilterState)
  ).toStrictEqual([todos[1]])
})

it('shoudl return active todos', () => {
  const visibilityFilterState: VisibilityFilterState = {
    visibilityFilter: 'SHOW_ACTIVE',
  }
  expect(
    todoSelector.resultFunc(todoState, visibilityFilterState)
  ).toStrictEqual([todos[0], todos[2]])
})
