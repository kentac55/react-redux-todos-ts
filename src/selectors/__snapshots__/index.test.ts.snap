// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return application state 1`] = `
Object {
  "error": null,
  "loaded": false,
  "loading": false,
}
`;

exports[`should return filter state 1`] = `"SHOW_ALL"`;

exports[`should return todo state 1`] = `Array []`;
